proposals/mmmm-spawn-let.md

```diff
@@ -0,0 +1,701 @@
+
+```swift
+func go2() async {
+  spawn let f = fast()
+  spawn let s = slow()
+  _ = await f
+  return "nevermind..."
+  // implicitly: cancels s
+  // implicitly: awaits s
+}
+```
+
+The duration of the `go2()` call remains the same, it is always `time(go2) == max(time(f), time(s))`.
+
+Special attention needs to be given to the `spawn let _ = ...` form of declarations. This form is interesting because it creates a child-task of the right-hand-side initializer, however it actively chooses to ignore the result. Such a declaration (and the associated child-task) will run and be awaited-on implicitly, as the scope it was declared in is about to exit — the same way as an unused `spawn let` declaration would be.
+
+> It may be interesting for a future proposal to explore the viability of sugar to `spawn voidReturningFunction()` directly, as a `Void` returning function may often not necessarily want to be awaited on for control-flow reasons, as variables of interesting types would be.
+
+### `spawn let` and closures
+
+Because `spawn let` tasks cannot out-live the scope in which they are defined, passing them to closures needs some further discussion for what is legal and not.
+
+It is legal to capture a `spawn let` in a non-escaping asynchronous closure, like this:
+
+```swift
+func greet(_ f: () async -> String) async -> String { await f() }
+
+spawn let name = "Alice"
+await greet { await name }
+```
+
+Notice how we are required to write the `await` inside the closure as well as in front of the `greet` function. This is on purpose as we do want to be explicit about the await inside the closure. 
+
+The same applies to auto closures, in order to make it explicit that the await is happening _inside_ the closure rather than before it, it is required to await explicitly in parameter position where the auto closure is formed for the argument:
+
+```swift
+func greet(_ f: @autoclosure () async -> String) async -> String { await f() }
+
+spawn let name = "Bob"
+await greet(await name) // await on name is required, because autoclosure
+```
+
+It is *not* legal to escape a `spawn let` value to an escaping closure. This is because structures backing the spawn let implementation may be allocated on the stack rather than the heap. This makes them very efficient, and makes great use of the structured guarantees they have to adhere to. These optimizations, however, make it unsafe to pass them to any escaping contexts:
+
+```swift
+func greet(_ f: @escaping () async -> String) async -> String { somewhere = f; somewhere() }
+
+spawn let name = "Bob"
+await greet { await name } // error: cannot escape 'spawn let' value
+```
+
+
+
+> Note: If Swift had a `@useImmediately` annotation that could be used together with even escaping closures, as they would "promise" to be called immediately without detaching or storing the closure elsewhere.
```

Comment:
I'm not sure that this is true, because `detach` itself could be considered to `@useImmediately` its closure argument. To avoid lifetime-extending the `async let` task, any potential awaits of the value have to happen before it goes out of scope, which seems to me like it really does require the closure be nonescaping. We could however potentially allow for escaping closures to capture the value by awaiting it, with syntax like `{[await name] in name}`, perhaps.

Reply:
Ah I think you're right on that one, we cannot indefinitely extend the lifetime of the let after all... Thanks Joe.

I'll remove the note and add a future direction about capturing with the proposed `[await x]` capture list 👍